services/property-svc/internal/property/aggregate/aggregate.go

```diff
@@ -101,3 +110,115 @@ func (a *PropertyAggregate) onSetIDUpdated(evt hwes.Event) error {
 	a.Property.SetID = setID
 	return nil
 }
+
+func (a *PropertyAggregate) onSubjectTypeUpdated(evt hwes.Event) error {
+	var payload propertyEventsV1.PropertySubjectTypeUpdatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	value, found := pb.SubjectType_value[payload.SubjectType]
+	if !found {
+		return fmt.Errorf("invalid subjectType: %s", payload.SubjectType)
+	}
+	subjectType := (pb.SubjectType)(value)
+
+	a.Property.SubjectType = subjectType
+	return nil
+}
+
+func (a *PropertyAggregate) onFieldTypeUpdated(evt hwes.Event) error {
+	var payload propertyEventsV1.PropertyFieldTypeUpdatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	value, found := pb.FieldType_value[payload.FieldType]
+	if !found {
+		return fmt.Errorf("invalid fieldType: %s", payload.FieldType)
+	}
+	fieldType := (pb.FieldType)(value)
+
+	a.Property.FieldType = fieldType
+	return nil
+}
+
+func (a *PropertyAggregate) onNameUpdated(evt hwes.Event) error {
+	var payload propertyEventsV1.PropertyNameUpdatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	a.Property.Name = payload.Name
+	return nil
+}
+
+func (a *PropertyAggregate) onAllowFreetextUpdated(evt hwes.Event) error {
+	var payload propertyEventsV1.FieldTypeDataAllowFreetextUpdatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	if a.Property.FieldTypeData.SelectData != nil {
+		a.Property.FieldTypeData.SelectData.AllowFreetext = payload.NewAllowFreetext
+	}
+
+	return nil
+}
+
+func (a *PropertyAggregate) onFieldTypeNoneUpdated(evt hwes.Event) error {
+	var payload propertyEventsV1.FieldTypeDataNoneUpdatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	a.Property.FieldTypeData = models.FieldTypeData{None: &payload.None}
+	return nil
+}
+
+func (a *PropertyAggregate) onFieldTypeDataSelectOptionsUpserted(evt hwes.Event) error {
+	var payload propertyEventsV1.FieldTypeDataSelectOptionsUpsertedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	if a.Property.FieldTypeData.SelectData != nil {
```

Comment:
Here we have one of those race-conditions, which I was worried will happen when we event source too much. Say we want to change a Properties Field Type from Text to Select. Then we emit both "UpdateFieldType" and, "UpsertSelectOptions" events. The order of which is not guaranteed iirc (@MaxSchaefer right?). This means if this event gets processed first, this if is false, and the event silently dropped. If the other event makes it first, we actually commit the new options.

Reply:
If we actually face these issues I could just break down all the Events related to the `Property.FieldTypeData` down to something like `PropertyFieldTypeDataUpdatedEvent`

Reply:
The order is which the events are applied to the aggregate is guaranteed.
We are talked about the optimistic concurrency. If another aggregate gets loaded and committed to the event store while another request is in flight, the request that commits with the now outdated aggregate version gets rejected.
 
![Bildschirmfoto 2024-03-14 um 12 12 07 AM](https://github.com/helpwave/services/assets/40192999/9dcc5cba-7402-4124-b0c5-bbf84bddbadd)